test/distributed/tensor/parallel/test_tp_examples.py

```diff
@@ -282,18 +291,48 @@ def test_transformer_training(self, is_seq_parallel=False):
 
         # Compare outputs on the same input.
         output = model(inp)
-        output_tp = model_tp(inp)
+        with CommDebugMode() as comm_mode:
+            output_tp = model_tp(inp)
         self.assertEqual(output, output_tp)
+        if is_seq_parallel:
+            self.assertDictEqual(comm_mode.get_comm_counts(), {
+                c10d_functional.all_reduce: 1,
+                c10d_functional.reduce_scatter_tensor: 4,
+                c10d_functional.all_gather_into_tensor: 7,
+            })
+        else:
+            self.assertDictEqual(comm_mode.get_comm_counts(), {
+                c10d_functional.all_reduce: 5,
+                c10d_functional.all_gather_into_tensor: 2,
+            })
 
         # Ensure gradients are equal.
         output.sum().backward()
-        output_tp.sum().backward()
+        with CommDebugMode() as comm_mode:
+            output_tp.sum().backward()
         self._check_module(model, model_tp, check_grad=True)
+        if is_seq_parallel:
+            self.assertDictEqual(comm_mode.get_comm_counts(), {
+                c10d_functional.reduce_scatter_tensor: 4,
+                c10d_functional.all_gather_into_tensor: 7,
+            })
+        else:
+            self.assertDictEqual(comm_mode.get_comm_counts(), {
+                c10d_functional.all_reduce: 8,
+                c10d_functional.all_gather_into_tensor: 1,
+            })
 
         # Ensure model weights are still the same after update.
         optim.step()
-        optim_tp.step()
+        with CommDebugMode() as comm_mode:
+            optim_tp.step()
         self._check_module(model, model_tp)
+        if is_seq_parallel:
+            self.assertDictEqual(comm_mode.get_comm_counts(), {
+                c10d_functional.all_reduce: 30,
```

Comment:
wat, I am shocked that there're 30 allreduces in optimizer step 😮 I was expecting around 5 allreduces hmm

could you pass `foreach=True` to Adam optimizer and see if that would reduce # allreduces?

Reply:
I think foreach is a single kernel but the tensors are still kept separate

Reply:
It seems each LayerNorm would incur 2 all-reduces, weight & bias. With default two-layer Transformer, we have two `attention_norm`, two `ffn_norm`, and one final `norm` -- adding up to 10 all-reduces already?

In general, I wonder if we should include a breakdown of the collectives in the test -- ideally it should be a function on the number of TransformerBlocks (and other relevant configs), so even if we modify the model architecture, the test could still pass.

Reply:
```
Traceback (most recent call last):
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 540, in wrapper
    self._join_processes(fn)
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 759, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 809, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 656, in run_test
    getattr(self, test_name)()
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 542, in wrapper
    fn()
  File "/home/tristanr/pytorch/torch/testing/_internal/common_utils.py", line 2739, in wrapper
    method(*args, **kwargs)
  File "/home/tristanr/pytorch/torch/testing/_internal/common_utils.py", line 439, in instantiated_test
    test(self, **param_kwargs)
  File "/home/tristanr/pytorch/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 374, in wrapper
    func(self, *args, **kwargs)  # type: ignore[misc]
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 181, in wrapper
    return func(*args, **kwargs)
  File "/home/tristanr/pytorch/test/distributed/tensor/parallel/test_tp_examples.py", line 247, in test_transformer_training
    optim_tp.step()
  File "/home/tristanr/pytorch/torch/optim/optimizer.py", line 391, in wrapper
    out = func(*args, **kwargs)
  File "/home/tristanr/pytorch/torch/optim/optimizer.py", line 76, in _use_grad
    ret = func(self, *args, **kwargs)
  File "/home/tristanr/pytorch/torch/optim/adam.py", line 168, in step
    adam(
  File "/home/tristanr/pytorch/torch/optim/adam.py", line 318, in adam
    func(params,
  File "/home/tristanr/pytorch/torch/optim/adam.py", line 522, in _multi_tensor_adam
    torch._foreach_lerp_(device_exp_avgs, device_grads, 1 - beta1)
  File "/home/tristanr/pytorch/torch/distributed/_tensor/api.py", line 279, in __torch_dispatch__
    return DTensor._op_dispatcher.dispatch(
  File "/home/tristanr/pytorch/torch/distributed/_tensor/dispatch.py", line 111, in dispatch
    op_info = self.unwrap_to_op_info(op_call, args, kwargs)
  File "/home/tristanr/pytorch/torch/distributed/_tensor/dispatch.py", line 314, in unwrap_to_op_info
    raise RuntimeError(
RuntimeError: aten._foreach_lerp_.Scalar: got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!
```

Seems to error out with `foreach=True`

Reply:
If you want to bypass that error:
```
from torch.distributed._tensor.experimental import implicit_replication

with implicit_replication():
    optim.step()
```
cc: @wanchaol

Reply:
Ohh i see, this is nn.Layernorm not RMSNorm, where it have 2 allreduces per-layernorm, this make sense then